### ISSUE TITLE
multiple ssl backends don't work as expected

### ISSUE THREAD

**Issue by notjaywu** [curl/curl#12663]
### I did this

I built curl locally by following https://curl.se/docs/install.html

I have the following variables for both cases:
```
export ARCH=arm64
export SDK=macosx
export DEPLOYMENT_TARGET=10.15
export CFLAGS="-arch $ARCH -isysroot $(xcrun -sdk $SDK --show-sdk-path) -m$SDK-version-min=$DEPLOYMENT_TARGET"
```

## Case 1
* curl is built with `./configure --host=$ARCH-apple-darwin --prefix=$(pwd)/artifacts --with-secure-transport`
* and run `./src/curl -v https://localhost:8005`
 
I can receive the response correctly from the local TLS server.
```
* Host localhost:8005 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8005...
* Connected to localhost (::1) port 8005
* ALPN: curl offers http/1.1
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: **@**.local (**)
* Server certificate: mkcert **@**.local (**)
* using HTTP/1.x
> GET / HTTP/1.1
> Host: localhost:8005
> User-Agent: curl/8.5.0-DEV
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Tue, 09 Jan 2024 07:57:11 GMT
< Connection: keep-alive
< Keep-Alive: timeout=5
< Transfer-Encoding: chunked
<
hello world
* Connection #0 to host localhost left intact
```

## Case 2
* curl is built with `./configure --host=$ARCH-apple-darwin --prefix=$(pwd)/artifacts --with-ssl=/opt/homebrew/opt/openssl@3.0 --with-secure-transport --with-default-ssl-backend=openssl`
* run `CURL_SSL_BACKEND=secure-transport ./src/curl -v https://localhost:8005`
 
I will get the following error:
```
* Host localhost:8005 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8005...
* Connected to localhost (::1) port 8005
* ALPN: curl offers http/1.1
* SSL: peer not verified:  RecoverableTrustFailure
* Closing connection
curl: (60) SSL: peer not verified:  RecoverableTrustFailure
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

### I expected the following

Curl built with multiple ssl backends enabled in Case 2 can get the response correctly.

### curl/libcurl version

curl 8.5.0-DEV (aarch64-apple-darwin) libcurl/8.5.0-DEV (SecureTransport) OpenSSL/3.0.12 zlib/1.2.11
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTPS-proxy IPv6 Largefile libz MultiSSL NTLM SSL threadsafe TLS-SRP UnixSockets

### operating system

Darwin xx.local 21.4.0 Darwin Kernel Version 21.4.0: Mon Feb 21 20:35:58 PST 2022; root:xnu-8020.101.4~2/RELEASE_ARM64_T6000 arm64

**Comment by jay** [comment]
> * SSL: peer not verified:  RecoverableTrustFailure

that is a secure transport error [kSecTrustResultRecoverableTrustFailure](https://developer.apple.com/documentation/security/sectrustresulttype/ksectrustresultrecoverabletrustfailure?language=objc), not an openssl error. are you sure you're using the curl built for both? use `curl -V` to check. the inactive SSL backends should be in parentheses.

also ... homebrew apparently doesn't update `/etc/ssl/cert.pem` for openssl (which has since for some reason disappeared from your verbose output), so that may be something? https://github.com/orgs/Homebrew/discussions/2444

**Comment by bagder** [comment]
> that is a secure transport error

I think that is the point. In the first case with a single TLS backend it works, but when you do two backends and explicitly select the same one as in the first case (Secure Transport), it fails.

**Comment by notjaywu** [comment]
Thanks for your quick response!

Yes, I am sure the curl is built for both.

## 1
```
./src/curl --version

curl 8.5.0-DEV (aarch64-apple-darwin) libcurl/8.5.0-DEV (SecureTransport) OpenSSL/3.0.12 zlib/1.2.11
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTPS-proxy IPv6 Largefile libz MultiSSL NTLM SSL threadsafe TLS-SRP UnixSockets
```

##2
```
CURL_SSL_BACKEND=secure-transport ./src/curl --version

curl 8.5.0-DEV (aarch64-apple-darwin) libcurl/8.5.0-DEV SecureTransport (OpenSSL/3.0.12) zlib/1.2.11
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTPS-proxy IPv6 Largefile libz MultiSSL NTLM SSL threadsafe TLS-SRP UnixSockets
```

**Comment by notjaywu** [comment]
@jay 

I think it may not be related to the `/etc/ssl/cert.pem` file of openssl.

I just want to access the keychain on macOS to verify the server. With a single `SecureTransport` backend, it can access the keychain correctly, but in multiple-ssl-backends mode, it fails.

**Comment by bagder** [comment]
I suspect the difference is this: configure checks for and adds a define for `CURL_CA_BUNDLE` only if one of a subset of TLS libraries is selected. OpenSSL being one of them. Secure transport is not one of them.

When you build with both libs enabled, you get the path set set but without OpenSSL you do not get it set. So the `verify_cert` function does not get called by default without OpenSSL in the build because `(conn_config->CAfile || conn_config->ca_info_blob)` equals false.

**Comment by notjaywu** [comment]
I guess self-signed certificates always fail in the `verify_cert()` function.

As I tested, the difference is:
The single SecureTranport endpoint will not return a `-9841` code, so it doesn't call the `verify_cert()` function at all.
But in multiple SSL endpoints mode, the SecureTransport returns at least one `-9841` code.

**Comment by bagder** [comment]
> I guess self-signed certificates always fail in the verify_cert() function.

Unless you have the signing CA in the bundle used for verification, which is how that works.

**Comment by bagder** [comment]
>  The single SecureTranport endpoint will not return a -9841 code, so it doesn't call the verify_cert() function at all.

Right, but I think it does because of the same reason: there is earlier code paths that enable the use of the CA bundle when that is set.

**Comment by bagder** [comment]
@notjaywu is this fixed now with #12704 merged?

**Comment by notjaywu** [comment]
> @notjaywu is this fixed now with #12704 merged?

Yes, it worked. Thank you!
